### DISCUSSION TITLE
Problem with multiple transparent textures

### DISCUSSION THREAD

**Discussion by B1azA** [gfx-rs/wgpu#3566] [Q&A] [answered]
Hi, I am trying to render two transparent textures, but I have problem with rendering order.

![bad](https://user-images.githubusercontent.com/77540263/223440632-c07913de-4cd7-410a-84cb-b7065b77adbf.png)
![ok](https://user-images.githubusercontent.com/77540263/223440637-1f8907a2-81ee-4f61-802a-1d8f71c9871c.png)

The A texture is behind the color texture.
If I render A texture first, everything is fine. However, if I do it the other way, the color texture will cover the A texture.

Is there any other way to fix this than changing the rendering order?
Thanks

**Comment by kpreid** [comment, marked as answer]
Correct transparency rendering, in the general case, is a fundamentally hard problem which GPUs cannot just solve for you implicitly.

In your case, it looks like you're interested in shapes that are either fully opaque or fully transparent; you can handle these by checking the alpha value in the fragment shader and running `discard;` if it is zero. This will stop the A being hidden in the first case. It will not help with semi-transparent antialiased edges; you can either remove all partial transparency from your input textures and live with jagged edges, or use signed distance field (SDF) rendering, which can make a smooth edge at any scale from a properly prepared texture. (This is particularly popular for text rendering.)

If you do want to deal with partial transparency, then you need to do one of these things:

* sort things from back to front
* disable depth testing, so that which thing appears in front is fully determined by the drawing order and not by depth
* use order-independent transparency (OIT); one of various schemes to get _approximately_ correct blending regardless of drawing order
